### ISSUE TITLE
[BUG] WebUI out of mem with a large number of instance variable definitions

### ISSUE THREAD

**Issue by thedist** [bitfocus/companion#2055]
### Is this a bug in companion itself or a module?

- [X] I believe this to be a bug in companion

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

While working on the Google Sheets module I have set up instance variables for each cell, such as `SpreadsheetName_sheet!A1`, and have found that while the number of instance variables scales quite well, their definitions do not.

I do not currently know the exact point at which this becomes an issue, but with my current testing in the Google Sheets module it has about 100,000 cells, but with this many definitions it causes the browser tab showing the WebUI to run out of memory and crash either when attempting to view the variables tab for the module, or when starting to type an instance variable for a button and the autocomplete starts loading the definitions list.

As this is just a limitation of Instance Variable Definitions, I think one solution to this issue may be to create a hard limit within Companion on the number of definitions a module can register by simply slicing the array of definitions sent by the module to some limit (and maybe logging a warning). This way regardless of what a module attempts to do Companion wont let it negatively impact the WebUI.

### Steps To Reproduce

1. Create a module
2. Use a loop to create a large number of Instance Variable Definitions
3. Try browse the instance variables, or start typing one on a button, and monitor the tabs mem usage

### Expected Behavior

Expected behaviour would be for instance variable definitions to be limited so that at no point can a module break the WebUI (or any core part of Companion)

### Environment (please complete the following information)

```markdown
- OS: Win 10
- Browser: Chrome
- Companion Version: 2.2.2 beta
```

**Comment by Julusian** [comment]
I expect that the point it becomes an issue will vary depending on the machine viewing the webui. And potentially also the network between the two, as a large number of variables will be a large blob of data to transfer when loading the page.

It would be interesting to know where all the memory is going. Is it the size of the json, feeding that into tribute, the dropdowns which use variables or something else?

But yes a good starting point for this would be to put some limit on the number of variables that a module can expose

**Comment by Julusian** [comment]
Alternatively, we could look at whether we could do more with dynamic loading of the variables instead.
We could likely get away with not loading the full list of variable-definitions to the client, and instead doing queries to the backend as the user types. It might make the ui feel a bit slower when doing this, but equally I have noticed that the dropdowns are noticably laggy at times

**Comment by thedist** [comment]
Could a hybrid approach of sorts work? Rather than doing queries each letter to the backend, or having the entire instance variable list on the client, we could instead have it so as you start typing `${` it'll just use the list of instances to disable all of the instance labels, eg `${internal:`, `${http:`, `${vmix:` etc... and when you type one of the labels it'll then make a request to the backend to for the instance variables of that instance.

The benefit of this is that it'd limit it to only ever trying to display, and filter as you type, just the instance variables for 1 instance. Should 1 instance have an insane amount of variables it wont break trying to use variables from any other instance, and should still have the responsiveness of filtering as the user types client-side.

The downside to this would be no instance variable suggestions until a user starts typing.

Additionally, one way to perhaps further help would be to add a `category` property to instance variables. Much like when browsing presets they are not only divided into instances, but each module can categorise presets. Same thing could be done with instance variables, and for backwards compatibility any instance variable without a category could be always shown rather than nested in a category.

**Comment by Julusian** [comment]
With a couple of simple ui changes I am able to run with 750,000 definitions. Things are definitely a bit slow, and sometimes loading the ui does timeout (even when connecting over localhost).  
Companion itself is perfectly fine with this number, and is running at around 700mb of memory (I didnt give them values, just the definitions). It looks to be pushing ~80mb of json to the ui for this.

80mb is a lot of data to be pushing to the ui, but it only really becomes problematic when we try and draw them all at once in the DOM. My simple fix has been to cap the number that we will use at once:
![image](https://user-images.githubusercontent.com/1327476/202039260-b9d41536-debb-4590-94fc-70db4e753836.png).

I have found that even without this ridiculous number of variables, that the text input with variables suggestions feels rather sluggish. It might be time to start properly investigating replacing tributejs with something more react friendly (https://www.npmjs.com/package/react-autocomplete-input looks reasonable).
Trying to use a variable in the internal feedbacks which present a dropdown is causing chrome the chrome tab to be unresponsive.

I've been thinking more on whether we can avoid loading the full list of variable definitions into the ui, but I have thought of some more complications:
* I really like that it is possible to find a variable by searching for a keyword, you don't have to fill in the instance label first
* Anywhere that a variable is used may need to load in the variable definitions for any instance (internal actions/feedbacks/triggers)
* The definitions will need to be cached for a little while, to avoid churn of unloading them briefly. (This should be fine to do)

With all that, I'm not sure what the best solution here is.

**Comment by thedist** [comment]
I wonder how responsive it would be to offload autocomplete entirely to the Backend.

For example, as you start typing if the string includes the start of an Instance variable `$(` with no closing paren to end it, the element could send the current text string through the websocket to Companion under an 'autocomplete' event. Companion could then check through the instance variables as it would always have all of them loaded, and then respond with 5 to 10 results (and an indicator that there's x number of results not shown). As the users enters more characters it'll progressively narrow down the results. This will mean that the frontend wont need to have all instance variables loaded.

The only time the frontend would need to work with a large set of instance variables would be when going to the variables page, but this again could be broken down into smaller data sets by only requesting from Companion just the variables for the category, eg 'Custom Variables', or for one of the modules. For modules with very large number of variables (Google Sheets) we could add a 'category' value to instance variables, so that when I go to the variable page for Google Sheets, rather than having hundreds of thousands of variables, I would instead get another set of buttons to go to one of the specific categories, with all variables lacking a category going to a 'default' page to allow for backwards compatibility for all existing modules.

**Comment by Julusian** [comment]
@thedist I'm curious if beta 9339 is good enough on this now?
I haven't touched the text input suggestions yet, everywhere else should be using the new virtualising (I will start on that input soon). For a representitive input field, try the 'Variable: Check value' feedback

Since this was originally opened the variables dropdown has gone from tribute -> react-select -> baseui (with tanstack virtual), so behaviour is very likely different. Even with only 8612 variables from a module it feels much smoother.

It could for sure still be more efficient on bandwidth if lazily loading, but if this is good enough it keeps it simpler

**Comment by thedist** [comment]
So I just did a quick test running 9343 with 50,399 variables.

The Variables page is quite performant, there is a perceivable initial loading delay about about 1.5s but I think this is reasonable and most importantly it doesn't crash the browser by trying to load all of those variables as elements in the DOM. Search works, as does scrolling, and it's noticeable that it's working with a large set of variables but well within being useable.

On the bandwidth side of things, while on the Variables page for the connection it's querying for variables once a second, and getting over 2MB of data for ~50k variables. Even if there's no changes, causing CPU usage to spike to about 30 each time. This was still usable for me, but for slower machines/mobile devices it may be more problematic. I also didn't test with more extreme numbers of variables. I think we still need the full variable list for responsive searching and quickly scrolling, but maybe this could be changed so the initial query is a full set of variable and then the polling is just for changes?

The only other suggestion I have here is on the main Variables page, clicking the connection with ~50k variables has a little bit of a load time before it loads that list of connection variables, and alternative to this might be to immediately change page (eg, going from `/variables` to `/variables/connection/test`) render a skeleton of the variables list, and then fill in that skeleton once ready. Could also do something like a spinner but that would be annoying on variable lists that are almost instant as you'd have a spinner flash up for a fraction of a second, where as skeletons are more user friendly. That being said though this is very much a low priority as it's purely cosmetic and minor UX improvement, the actual functionality when handling large numbers of variables is good.

For the `Variable: Check value` feedback as you suggested testing, again similar experience with the Variables page in that it didn't crash anything, both CPU and Mem usage was reasonable when typing, no real concerns here so it looks good to me!

I tested just a test input that supported variables for comparison, the memory footprint on the tab went from ~400MB to over 1.2GB, and each time I typed a character in to search Chrome was showing CPU spiking to about 120 (as opposed to 20 to 30, and for a shorter duration, when searching in the Variables list or the check value feedback). It didn't crash as it's not all being rendered in the DOM, and it was somewhat usable for me, but it was not a smooth experience and I would imagine could still crash the tab on slower machines/mobile devices. So if this is updated the same way the other places do variables then that would be amazing.

tl;dr, nothing crashed, and while there was a perceivable delay when loading/searching variable lists it was still very much usable and the performance was definitely reasonable for that number of variables.

**Comment by Julusian** [comment]
> The Variables page is quite performant, there is a perceivable initial loading delay about about 1.5s

In a dev or prod build? I've noticed that some routes are particularly slow in dev mode, but much faster in a production build. But maybe something that can be done

> I think we still need the full variable list for responsive searching and quickly scrolling, but maybe this could be changed so the initial query is a full set of variable and then the polling is just for changes?

That polling is specifically for the values, so that we don't have to stream all the values of every variable change into the ui. But we do now have the ability to subscribe to individual variable values, so I suppose we could try using that for just the values which are in the DOM. Might cause some delayed population of values while scrolling, but that is probably acceptable

 > I tested just a test input that supported variables for comparison, the memory footprint on the tab went from ~400MB to over 1.2GB, and each time I typed a character 

The weird thing is that this component is using virtualisation/windowing already, and has for a long time.. I dont know why it still struggles so much, especially when the new dropdown was so simple to put together (that part of the functionality was, getting all the input flows correct was the hard part). Im really glad I felt like looking into ditching react-select now!

----

Perhaps if we want to further reduce the memory impact of the variable definitions, we could take some inspiration from the templating I did for presets. Something to think about as a follow up

**Comment by thedist** [comment]
> In a dev or prod build? I've noticed that some routes are particularly slow in dev mode, but much faster in a production build. But maybe something that can be done

This was all done in dev, so I just re-run the tests in the latest build. It was pretty much the same. One additional thing I did check that I only just thought of, I checked entering a variable in a text input that was set to Expression mode and that had pretty great performance, similar to the dropdown if not better, but when in non-Expression mode (but still a text input that support variables) it was spiking to over 120 CPU and memory going to over 900MB.

> That polling is specifically for the values, so that we don't have to stream all the values of every variable change into the ui. But we do now have the ability to subscribe to individual variable values, so I suppose we could try using that for just the values which are in the DOM. Might cause some delayed population of values while scrolling, but that is probably acceptable

I can imagine that being a pain when scrolling through variables as it'd be spamming unsubscribe/subscribe messages. I wasn't sure if on Companions side we could cache variable changes, so that on the initial load of the Variables page it queries the full list as it currently does, but on subsequent polls it only requests the changes, so the UI would still only need to re-render if there's a change to one of the variables currently being rendered, but now it's not having to go through a potentially unlimited number of variables to find the ones currently being rendered in check for differences.

> Perhaps if we want to further reduce the memory impact of the variable definitions, we could take some inspiration from the templating I did for presets. Something to think about as a follow up

More performance is always a good goal to aim for, and I think the state we're in now where by it doesn't crash browser tabs it's very much in a usable state even with a large number of variables. Perhaps something else we can do is simply remind module devs the impact of variables and if they are generating tens/hundreds of thousands of variables then they need to be aware of the impact.

**Comment by Julusian** [comment]
9367 has the rebuilt textinput field. I havent checked performance myself, but its the same inners as the dropdown field.

> I can imagine that being a pain when scrolling through variables as it'd be spamming unsubscribe/subscribe messages.

yeah I tried it and it wasnt great. It was messy to handle the sub/unsub, a naive first implemnetation caused some react crashes due to too many state updates during renders, and it was sluggish to show updates.

But perhaps reworking the polling wouldnt be that bad. It could be done as a subscription, only invalidated when one a variable for the connection changes (tracking which ones changed in a set), and with a debounce of 1s or something to stop rate limit updates.
